docs/references/architecture/adr-119-dog-mempool-gossip.md

```diff
@@ -0,0 +1,287 @@
+- 2024-11-25: Applied comments from @hvanz and added information on decided metrics (@jmalicevic)
+
+## Status
+Accepted: [#3297][tracking-issue].
+
+
+## Context
+
+The current transaction dissemination protocol of the mempool sends each received 
+transaction to all the peers a node is connected to, except for the sender. 
+While resilient to Byzantine attacks, this type of transaction gossiping is 
+causing a lot of network traffic and nodes receiving duplicate transactions
+very frequently. 
+
+Benchmarks have also confirmed a large portion of the sent and received bytes 
+is due to transaction gossiping. 
+
+DOG is a protocol that aims to reduce the number of duplicate transactions sent
+while maintaining the resilience to attacks. 
+
+## Alternative Approaches
+
+The existing alternative approaches for transaction gossiping are:
+- FLOOD - the current transaction gossiping protocol in CometBFT
+- CAT Mempool - the transaction gossiping proposed by Celestia
+- Limiting number of peers to send a transaction to - Experimental protocol in CometBFT
+- Etherium's transaction gossiping protocol.
+
+
+### FLOOD - the current transaction gossiping protocol in CometBFT
+
+Currently, CometBFT gossips each transaction to all the connected peers, except the sender
+of the transaction. Thus, for every transaction, CometBFT keeps a list of senders to make sure
+it is not sent again to a node that has sent it. 
+This does not prevent a node to send the transaction to a node that has sent it to the 
+sender, thus creating a lot of duplicates. 
+
+This is the exact problem DOG was designed to tackle.
+
+
+
+## Decision
+
+The CometBFT team has decided to implement the protocol on top of the existing mempool
+(`flood` ) dissemination protocol, but make it optional. The protocol can be activated/deactivated
+via a new config flag `enable_dog_protocol`. More details on this in the sections below. 
+
+## Detailed Design
+
+We start the design section with a general description of how the algorithm works. 
+
+CometBFT caches received transactions and stores the IDs of all the peers that have sent it in a `senders` list. When a transaction enters the mempool,
+if it was already received it can be retrieved from the cache. When this happens a certain number of times, the receiving node notifies the sender to stop forwarding any transaction from the peer 
+that had originally sent the transaction. The exact point when the notification is sent is defined below as a redundancy threshold. (TODO LiNK)
+
+Namely, let's assume that transaction `tx_1` was sent by node `B` to node `C` which forwarded it to `A`. `A` already has the transaction and notifies node `C` about this. Node `C` will look up the senders of transaction `tx_1` and in the future disable transaction forwarding from node `B` to node `A`.
+
+In reality, if node `C` ha received `tx_1` from multiple senders (`X`,`Y`, `B`), it will pick the first one that has sent it and disable that route. 
```

Comment:
The first one, except the first one? I mean, the transaction must be received first, then duplicates are.

Reply:
At the moment it just gets `senders[0]` and disabls the route between `senders[0]->A`